### PR TITLE
add isolate margin api

### PR DESCRIPTION
**Pull request by NNagato**

### CODE REVIEW

File: common.go [hunk 1]
```diff
@@ -451,3 +472,46 @@ type TickerEntry struct {
 	AskPrice string `json:"askPrice"`
 	AskQty   string `json:"askQty"`
 }
+
+type IsolatedMarginAssetInfo struct {
+	BaseAsset struct {
```
Comment (halink0803) [inline common.go]: This should be separate as a new type like MarginAsset or something like that. BaseAsset and QuoteAsset both is a MarginAsset.

File: binance_margin.go [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+// KeepListenKeyAliveIsolatedMargin keep it alive
+func (bc *Client) KeepListenKeyAliveIsolatedMargin(listenKey string) error {
+	return bc.keepListenKeyAlive(listenKey, listenKeyTypeIsolatedMarginAPI)
+}
+
+type marginCommonResult struct {
+	TranID uint64 `json:"tranId"`
+}
+
+func (bc *Client) CrossMarginTransfer(asset string, amount string, spotToMargin bool) (uint64, *FwdData, error) {
```
Comment (halink0803) [inline binance_margin.go]: If func var share the same type, you should group them like CrossMarginTransfer(asset, amount string, spotToMargin bool)

File: binance_margin.go [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+		SignedRequest(bc.secretKey)
+	fwd, err := bc.doRequest(rr, &result)
+	if err != nil {
+		return 0, fwd, err
+	}
+
+	return result.TranID, fwd, err
+}
+
+func (bc *Client) CrossMarginBorrow(asset string, isIsolated bool, symbol string, amount string) (uint64, *FwdData, error) {
```
Comment (halink0803) [inline binance_margin.go]: should group func variable

File: binance_margin.go [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+	}
+	sr := rr.SignedRequest(bc.secretKey)
+	fwd, err := bc.doRequest(sr, &result)
+	if err != nil {
+		return 0, fwd, err
+	}
+	return result.TranID, fwd, err
+}
+
+func (bc *Client) CrossMarginRepay(asset string, isIsolated bool, symbol string, amount string) (uint64, *FwdData, error) {
```
Comment (halink0803) [inline binance_margin.go]: should group function variables

File: binance_margin.go [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+	}
+	rr := req.WithHeader(apiKeyHeader, bc.apiKey).SignedRequest(bc.secretKey)
+	fwd, err := bc.doRequest(rr, &result)
+	if err != nil {
+		return result, fwd, err
+	}
+	return result, fwd, err
+}
+
+func (bc *Client) GetMaxBorrowable(asset string, isolatedSymbol string) (MaxBorrowableResult, *FwdData, error) {
```
Comment (halink0803) [inline binance_margin.go]: group variables pls

**Review by halink0803 — CHANGES_REQUESTED** [review]
Pls update

**Comment by NNagato** [comment]
@halink0803 PR is updated in this commit https://github.com/KyberNetwork/binance-client/pull/3/commits/c1a9857d1bd7dd2cc62bc870e4853ac368d4c321, pls re-review

File: binance_margin.go [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+}
+
+type marginCommonResult struct {
+	TranID uint64 `json:"tranId"`
+}
+
+func (bc *Client) CrossMarginTransfer(asset, amount string, spotToMargin bool) (uint64, *FwdData, error) {
+	transType := ""
+	if spotToMargin {
+		transType = "1"
```
Comment (halink0803) [inline binance_margin.go:42]: this should be set as constant

File: binance_margin.go [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+	}
+	sr := rr.SignedRequest(bc.secretKey)
+	fwd, err := bc.doRequest(sr, &result)
+	if err != nil {
+		return 0, fwd, err
+	}
+	return result.TranID, fwd, err
+}
+
+func (bc *Client) CrossMarginRepay(asset, symbol, amount string, isIsolated bool) (uint64, *FwdData, error) {
```
Comment (halink0803) [inline binance_margin.go]: this function used for isolated repay also then should it be MarginRepay only?

File: binance_margin.go [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+	isISO := ""
+	if isIsolated {
+		isISO = "TRUE"
+	} else {
```
Comment (halink0803) [inline binance_margin.go]: this is bad practice, it should be 
```
isISO := "FALSE"
if isIsolated {
  isISO = "TRUE"
}
```
no need `else` part

File: binance_margin.go [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+	)
+	requestURL := fmt.Sprintf("%s/sapi/v1/margin/repay", apiBaseURL)
+	req, err := NewRequestBuilder(http.MethodPost, requestURL, nil)
+	if err != nil {
+		return 0, nil, err
+	}
+	rr := req.WithHeader(apiKeyHeader, bc.apiKey).
+		WithParam("asset", asset).
+		WithParam("amount", amount)
+	if isIsolated {
```
Comment (halink0803) [inline binance_margin.go:103]: Actually, you can set it is "TRUE" here is enough

File: binance_margin.go [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+	fwd, err := bc.doRequest(rr, &result)
+	if err != nil {
+		return 0, fwd, err
+	}
+
+	return result.TranID, fwd, err
+}
+
+func (bc *Client) CrossMarginBorrow(asset, symbol, amount string, isIsolated bool) (uint64, *FwdData, error) {
+	isISO := ""
```
Comment (halink0803) [inline binance_margin.go]: this could be set as "FALSE" at first place. Please look at my comment in CrossMarginRepay.

**Review by halink0803 — CHANGES_REQUESTED** [review]
Please add comment to all exported function

**Comment by halink0803** [comment]
@NNagato please check

File: binance_margin.go (hunk 6 above)
Comment (NNagato) [reply binance_margin.go:42]: this param is only being used by this API and determined by param ```spotToMargin```, in isolated margin Binance uses another type for this. So I don't think we need to define the constant values for this

**Review by NNagato — COMMENTED** [review]
@halink0803 pls help me to check this PR again

**Review by halink0803 — APPROVED** [review]
LGTM

**Pull request merged**